File: src/weaponsprocurement/gui/StockReviewTradeContext.java
package weaponsprocurement.gui;

import weaponsprocurement.core.SubmarketWeaponStock;
import weaponsprocurement.core.WeaponStockRecord;
import weaponsprocurement.core.WeaponStockSnapshot;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

final class StockReviewTradeContext {
    private final List<StockReviewPendingTrade> pendingTrades;
    private final StockReviewQuoteBook quoteBook;
    private final Map<String, Integer> netByItem = new HashMap<String, Integer>();
    private final Map<String, Integer> buyByItem = new HashMap<String, Integer>();
    private final Map<String, Integer> sellByItem = new HashMap<String, Integer>();
    private final Map<String, Integer> affordableCache = new HashMap<String, Integer>();
    private final StockReviewPortfolioQuote portfolioQuote;
    private final int totalCost;
    private final float totalCargoSpaceDelta;
    private final float credits;
    private final float cargoSpaceLeft;

    StockReviewTradeContext(WeaponStockSnapshot snapshot, List<StockReviewPendingTrade> pendingTrades) {
        this.pendingTrades = pendingTrades;
        this.quoteBook = new StockReviewQuoteBook(snapshot);
        if (pendingTrades != null) {
            for (int i = 0; i < pendingTrades.size(); i++) {
                StockReviewPendingTrade trade = pendingTrades.get(i);
                add(netByItem, trade.getItemKey(), trade.getQuantity());
                if (trade.getQuantity() > 0) {
                    add(buyByItem, trade.getItemKey(), trade.getQuantity());
                } else if (trade.getQuantity() < 0) {
                    add(sellByItem, trade.getItemKey(), -trade.getQuantity());
                }
            }
        }
        portfolioQuote = quoteBook.quotePortfolio(pendingTrades);
        totalCost = portfolioQuote.totalCost();
        totalCargoSpaceDelta = portfolioQuote.totalCargoSpaceDelta();
        credits = StockReviewPlayerCargo.currentCredits();
        cargoSpaceLeft = StockReviewPlayerCargo.currentCargoSpaceLeft();
    }

    int netQuantityForItem(String itemKey) {
        return get(netByItem, itemKey);
    }

    int pendingBuyQuantityForItem(String itemKey) {
        return get(buyByItem, itemKey);
    }

    int pendingSellQuantityForItem(String itemKey) {
        return get(sellByItem, itemKey);
    }

    int buyableRemaining(WeaponStockRecord record) {
        return Math.max(0, record.getBuyableCount() - pendingBuyQuantityForItem(record.getItemKey()));
    }

    int sellableRemaining(WeaponStockRecord record) {
        return Math.max(0, record.getPlayerCargoCount() - pendingSellQuantityForItem(record.getItemKey()));
    }

    int positiveAdjustmentRemaining(WeaponStockRecord record, int requestedQuantity) {
        int requested = Math.max(0, requestedQuantity);
        if (requested <= 0) {
            return 0;
        }
        int sellCancellation = Math.min(requested, pendingSellQuantityForItem(record.getItemKey()));
        int remainingRequest = requested - sellCancellation;
        if (remainingRequest <= 0) {
            return sellCancellation;
        }
        return sellCancellation + affordableBuyQuantity(record, null, remainingRequest);
    }

    int negativeAdjustmentRemaining(WeaponStockRecord record, int requestedQuantity) {
        int requested = Math.max(0, requestedQuantity);
        if (requested <= 0) {
            return 0;
        }
        int buyCancellation = Math.min(requested, pendingBuyQuantityForItem(record.getItemKey()));
        int remainingRequest = requested - buyCancellation;
        if (remainingRequest <= 0) {
            return buyCancellation;
        }
        return buyCancellation + Math.min(remainingRequest, sellableRemaining(record));
    }

    int buyNeededForSufficiency(WeaponStockRecord record) {
        return Math.max(0, record.getDesiredCount() - (record.getOwnedCount() + netQuantityForItem(record.getItemKey())));
    }

    int sellableUntilSufficient(WeaponStockRecord record) {
        int stockAfterPlan = record.getOwnedCount() + netQuantityForItem(record.getItemKey());
        int excess = Math.max(0, stockAfterPlan - record.getDesiredCount());
        return Math.min(excess, sellableRemaining(record));
    }

    int transactionCostForItem(String itemKey) {
        return portfolioQuote.costForItem(itemKey);
    }

    int transactionCostForLine(String itemKey, String submarketId) {
        return portfolioQuote.costForLine(itemKey, submarketId);
    }

    int unitPriceForItem(WeaponStockRecord record) {
        if (record == null) {
            return StockReviewQuoteBook.PRICE_UNAVAILABLE;
        }
        int unitCost = quoteBook.nextBuyUnitPriceAfterPlannedBuys(record, pendingBuyQuantityForItem(record.getItemKey()));
        return unitCost == Integer.MAX_VALUE ? quoteBook.sellUnitPrice(record.getItemKey()) : unitCost;
    }

    int deltaToSufficient(WeaponStockRecord record) {
        int targetNet = record.getDesiredCount() - record.getOwnedCount();
        int delta = targetNet - netQuantityForItem(record.getItemKey());
        if (delta > 0) {
            return positiveAdjustmentRemaining(record, delta);
        }
        if (delta < 0) {
            return -negativeAdjustmentRemaining(record, -delta);
        }
        return 0;
    }

    List<StockReviewSellerAllocation> sellerAllocations(StockReviewPendingTrade trade) {
        if (trade == null) {
            return StockReviewQuote.ZERO.getSellerAllocations();
        }
        return portfolioQuote.sellerAllocations(trade.getItemKey(), trade.getSubmarketId());
    }

    int totalCost() {
        return totalCost;
    }

    float totalCargoSpaceDelta() {
        return totalCargoSpaceDelta;
    }

    int totalMarkupPaid() {
        return portfolioQuote.totalMarkupPaid();
    }

    float averageBuyMultiplier() {
        return portfolioQuote.averageBuyMultiplier();
    }

    float credits() {
        return credits;
    }

    float cargoSpaceLeft() {
        return cargoSpaceLeft;
    }

    boolean canConfirm() {
        return totalCost != StockReviewQuoteBook.PRICE_UNAVAILABLE
                && totalCost <= credits
                && totalCargoSpaceDelta <= cargoSpaceLeft + 0.01f;
    }

    int affordableBuyQuantity(WeaponStockRecord record, String submarketId, int requestedQuantity) {
        int maxByStock = Math.min(Math.max(0, requestedQuantity), buyableRemaining(record));
        if (submarketId != null) {
            maxByStock = Math.min(maxByStock, submarketRemaining(record, submarketId));
        }
        if (maxByStock <= 0) {
            return 0;
        }
        String key = record.getItemKey() + "|" + (submarketId == null ? "" : submarketId) + "|" + maxByStock;
        Integer cached = affordableCache.get(key);
        if (cached != null) {
            return cached.intValue();
        }
        int low = 0;
        int high = maxByStock;
        while (low < high) {
            int candidate = (low + high + 1) / 2;
            if (canAffordAdjustment(record, submarketId, candidate)) {
                low = candidate;
            } else {
                high = candidate - 1;
            }
        }
        affordableCache.put(key, Integer.valueOf(low));
        return low;
    }

    private boolean canAffordAdjustment(WeaponStockRecord record, String submarketId, int quantity) {
        StockReviewPortfolioQuote adjusted = quoteBook.quotePortfolio(StockReviewTradePlanner.withAdjustment(
                pendingTrades, record.getItemKey(), submarketId, quantity));
        int adjustedCost = adjusted.totalCost();
        if (adjustedCost == StockReviewQuoteBook.PRICE_UNAVAILABLE) {
            return false;
        }
        if (adjustedCost > credits) {
            return false;
        }
        return adjusted.totalCargoSpaceDelta() <= cargoSpaceLeft + 0.01f;
    }

    private int submarketRemaining(WeaponStockRecord record, String submarketId) {
        int sourceCount = 0;
        for (int i = 0; i < record.getSubmarketStocks().size(); i++) {
            SubmarketWeaponStock stock = record.getSubmarketStocks().get(i);
            if ((submarketId.equals(stock.getSourceId()) || submarketId.equals(stock.getSubmarketId())) && stock.isPurchasable()) {
                sourceCount += stock.getCount();
            }
        }
        int pendingFromSource = 0;
        if (pendingTrades != null) {
            for (int i = 0; i < pendingTrades.size(); i++) {
                StockReviewPendingTrade trade = pendingTrades.get(i);
                if (trade.matches(record.getItemKey(), submarketId) && trade.getQuantity() > 0) {
                    pendingFromSource += trade.getQuantity();
                }
            }
        }
        return Math.max(0, sourceCount - pendingFromSource);
    }

    private static void add(Map<String, Integer> counts, String itemKey, int quantity) {
        if (itemKey == null || quantity == 0) {
            return;
        }
        counts.put(itemKey, Integer.valueOf(get(counts, itemKey) + quantity));
    }

    private static int get(Map<String, Integer> counts, String itemKey) {
        Integer value = counts.get(itemKey);
        return value == null ? 0 : value.intValue();
    }

}

File: src/main/kotlin/weaponsprocurement/ui/stockreview/trade/StockReviewTradeContext.kt
package weaponsprocurement.ui.stockreview.trade

import weaponsprocurement.ui.stockreview.actions.*
import weaponsprocurement.ui.stockreview.state.*
import weaponsprocurement.ui.stockreview.rows.*
import weaponsprocurement.ui.stockreview.tooltips.*
import weaponsprocurement.ui.stockreview.rendering.*
import weaponsprocurement.ui.stockreview.trade.*

import weaponsprocurement.ui.*

import weaponsprocurement.stock.item.SubmarketWeaponStock
import weaponsprocurement.stock.item.WeaponStockRecord
import weaponsprocurement.stock.item.WeaponStockSnapshot
import java.util.HashMap

class StockReviewTradeContext(
    snapshot: WeaponStockSnapshot?,
    private val pendingTrades: List<StockReviewPendingTrade>?,
) {
    private val quoteBook = StockReviewQuoteBook(snapshot)
    private val netByItem = HashMap<String, Int>()
    private val buyByItem = HashMap<String, Int>()
    private val sellByItem = HashMap<String, Int>()
    private val affordableCache = HashMap<String, Int>()
    private val portfolioQuote: StockReviewPortfolioQuote
    private val totalCostValue: Long
    private val totalCargoSpaceDeltaValue: Float
    private val creditsValue: Float
    private val cargoSpaceLeftValue: Float

    init {
        if (pendingTrades != null) {
            for (trade in pendingTrades) {
                add(netByItem, trade.itemKey, trade.quantity)
                if (trade.quantity > 0) {
                    add(buyByItem, trade.itemKey, trade.quantity)
                } else if (trade.quantity < 0) {
                    add(sellByItem, trade.itemKey, -trade.quantity)
                }
            }
        }
        portfolioQuote = quoteBook.quotePortfolio(pendingTrades)
        totalCostValue = portfolioQuote.totalCost()
        totalCargoSpaceDeltaValue = portfolioQuote.totalCargoSpaceDelta()
        creditsValue = StockReviewPlayerCargo.currentCredits()
        cargoSpaceLeftValue = StockReviewPlayerCargo.currentCargoSpaceLeft()
    }

    fun netQuantityForItem(itemKey: String?): Int = get(netByItem, itemKey)

    fun pendingBuyQuantityForItem(itemKey: String?): Int = get(buyByItem, itemKey)

    fun pendingSellQuantityForItem(itemKey: String?): Int = get(sellByItem, itemKey)

    fun buyableRemaining(record: WeaponStockRecord): Int {
        return Math.max(0, record.buyableCount - pendingBuyQuantityForItem(record.itemKey))
    }

    fun sellableRemaining(record: WeaponStockRecord): Int {
        return Math.max(0, record.playerCargoCount - pendingSellQuantityForItem(record.itemKey))
    }

    fun positiveAdjustmentRemaining(record: WeaponStockRecord, requestedQuantity: Int): Int {
        val requested = Math.max(0, requestedQuantity)
        if (requested <= 0) return 0
        val sellCancellation = Math.min(requested, pendingSellQuantityForItem(record.itemKey))
        val remainingRequest = requested - sellCancellation
        if (remainingRequest <= 0) return sellCancellation
        return sellCancellation + affordableBuyQuantity(record, null, remainingRequest)
    }

    fun negativeAdjustmentRemaining(record: WeaponStockRecord, requestedQuantity: Int): Int {
        val requested = Math.max(0, requestedQuantity)
        if (requested <= 0) return 0
        val buyCancellation = Math.min(requested, pendingBuyQuantityForItem(record.itemKey))
        val remainingRequest = requested - buyCancellation
        if (remainingRequest <= 0) return buyCancellation
        return buyCancellation + Math.min(remainingRequest, sellableRemaining(record))
    }

    fun buyNeededForSufficiency(record: WeaponStockRecord): Int {
        return Math.max(0, record.desiredCount - (record.ownedCount + netQuantityForItem(record.itemKey)))
    }

    fun sellableUntilSufficient(record: WeaponStockRecord): Int {
        val stockAfterPlan = record.ownedCount + netQuantityForItem(record.itemKey)
        val excess = Math.max(0, stockAfterPlan - record.desiredCount)
        return Math.min(excess, sellableRemaining(record))
    }

    fun transactionCostForItem(itemKey: String?): Long = portfolioQuote.costForItem(itemKey)

    fun transactionCostForLine(itemKey: String?, submarketId: String?): Long {
        return portfolioQuote.costForLine(itemKey, submarketId)
    }

    fun unitPriceForItem(record: WeaponStockRecord?): Int {
        if (record == null) return StockReviewQuoteBook.PRICE_UNAVAILABLE
        val unitCost = quoteBook.nextBuyUnitPriceAfterPlannedBuys(record, pendingBuyQuantityForItem(record.itemKey))
        return if (unitCost == Int.MAX_VALUE) quoteBook.sellUnitPrice(record.itemKey) else unitCost
    }

    fun deltaToSufficient(record: WeaponStockRecord): Int {
        val targetNet = record.desiredCount - record.ownedCount
        val delta = targetNet - netQuantityForItem(record.itemKey)
        if (delta > 0) return positiveAdjustmentRemaining(record, delta)
        if (delta < 0) return -negativeAdjustmentRemaining(record, -delta)
        return 0
    }

    fun sellerAllocations(trade: StockReviewPendingTrade?): List<StockReviewSellerAllocation> {
        if (trade == null) return StockReviewQuote.ZERO.sellerAllocations
        return portfolioQuote.sellerAllocations(trade.itemKey, trade.submarketId)
    }

    fun totalCost(): Long = totalCostValue

    fun totalCargoSpaceDelta(): Float = totalCargoSpaceDeltaValue

    fun totalMarkupPaid(): Long = portfolioQuote.totalMarkupPaid()

    fun averageBuyMultiplier(): Float = portfolioQuote.averageBuyMultiplier()

    fun credits(): Float = creditsValue

    fun cargoSpaceLeft(): Float = cargoSpaceLeftValue

    fun canConfirm(): Boolean {
        return totalCostValue != StockReviewQuoteBook.PRICE_UNAVAILABLE.toLong() &&
            totalCostValue <= creditsValue &&
            totalCargoSpaceDeltaValue <= cargoSpaceLeftValue + 0.01f
    }

    fun affordableBuyQuantity(record: WeaponStockRecord, submarketId: String?, requestedQuantity: Int): Int {
        var maxByStock = Math.min(Math.max(0, requestedQuantity), buyableRemaining(record))
        if (submarketId != null) {
            maxByStock = Math.min(maxByStock, submarketRemaining(record, submarketId))
        }
        if (maxByStock <= 0) return 0

        val key = record.itemKey + "|" + (submarketId ?: "") + "|" + maxByStock
        val cached = affordableCache[key]
        if (cached != null) return cached

        var low = 0
        var high = maxByStock
        while (low < high) {
            val candidate = (low + high + 1) / 2
            if (canAffordAdjustment(record, submarketId, candidate)) {
                low = candidate
            } else {
                high = candidate - 1
            }
        }
        affordableCache[key] = low
        return low
    }

    private fun canAffordAdjustment(record: WeaponStockRecord, submarketId: String?, quantity: Int): Boolean {
        val adjusted = quoteBook.quotePortfolio(
            StockReviewTradePlanner.withAdjustment(pendingTrades, record.itemKey, submarketId, quantity)
        )
        val adjustedCost = adjusted.totalCost()
        if (adjustedCost == StockReviewQuoteBook.PRICE_UNAVAILABLE.toLong()) return false
        if (adjustedCost > creditsValue) return false
        return adjusted.totalCargoSpaceDelta() <= cargoSpaceLeftValue + 0.01f
    }

    private fun submarketRemaining(record: WeaponStockRecord, submarketId: String): Int {
        var sourceCount = 0
        for (stock: SubmarketWeaponStock in record.submarketStocks) {
            if ((submarketId == stock.sourceId || submarketId == stock.submarketId) && stock.isPurchasable()) {
                sourceCount += stock.count
            }
        }

        var pendingFromSource = 0
        if (pendingTrades != null) {
            for (trade in pendingTrades) {
                if (trade.matches(record.itemKey, submarketId) && trade.quantity > 0) {
                    pendingFromSource += trade.quantity
                }
            }
        }
        return Math.max(0, sourceCount - pendingFromSource)
    }

    companion object {
        private fun add(counts: MutableMap<String, Int>, itemKey: String?, quantity: Int) {
            if (itemKey == null || quantity == 0) return
            counts[itemKey] = get(counts, itemKey) + quantity
        }

        private fun get(counts: Map<String, Int>, itemKey: String?): Int {
            return counts[itemKey] ?: 0
        }
    }
}